server/processors/TurnOrderProcessor.cpp

```diff
@@ -17,18 +17,134 @@
 
 #include "../../lib/CPlayerState.h"
 #include "../../lib/NetPacks.h"
+#include "../../lib/pathfinder/CPathfinder.h"
+#include "../../lib/pathfinder/PathfinderOptions.h"
 
 TurnOrderProcessor::TurnOrderProcessor(CGameHandler * owner):
 	gameHandler(owner)
 {
 
 }
 
-bool TurnOrderProcessor::canActSimultaneously(PlayerColor active, PlayerColor waiting) const
+int TurnOrderProcessor::simturnsTurnsMaxLimit() const
 {
+	return gameHandler->getStartInfo()->simturnsInfo.optionalTurns;
+}
+
+int TurnOrderProcessor::simturnsTurnsMinLimit() const
+{
+	return gameHandler->getStartInfo()->simturnsInfo.requiredTurns;
+}
+
+void TurnOrderProcessor::updateContactStatus()
+{
+	blockedContacts.clear();
+
+	assert(actedPlayers.empty());
+	assert(actingPlayers.empty());
+
+	for (auto left : awaitingPlayers)
+	{
+		for(auto right : awaitingPlayers)
+		{
+			if (left == right)
+				continue;
+
+			if (computeCanActSimultaneously(left, right))
+				blockedContacts.push_back({left, right});
+		}
+	}
+}
+
+bool TurnOrderProcessor::playersInContact(PlayerColor left, PlayerColor right) const
+{
+	// TODO: refactor, cleanup and optimize
+
+	boost::multi_array<bool, 3> leftReachability;
+	boost::multi_array<bool, 3> rightReachability;
+
+	int3 mapSize = gameHandler->getMapSize();
+
+	leftReachability.resize(boost::extents[mapSize.z][mapSize.x][mapSize.y]);
+	rightReachability.resize(boost::extents[mapSize.z][mapSize.x][mapSize.y]);
+
+	const auto * leftInfo = gameHandler->getPlayerState(left, false);
+	const auto * rightInfo = gameHandler->getPlayerState(right, false);
+
+	for(const auto & hero : leftInfo->heroes)
+	{
+		CPathsInfo out(mapSize, hero);
+		auto config = std::make_shared<SingleHeroPathfinderConfig>(out, gameHandler->gameState(), hero);
+		CPathfinder pathfinder(gameHandler->gameState(), config);
+		pathfinder.calculatePaths();
+
+		for (int z = 0; z < mapSize.z; ++z)
+			for (int y = 0; y < mapSize.y; ++y)
+				for (int x = 0; x < mapSize.x; ++x)
+					if (out.getNode({x,y,z})->reachable())
+						leftReachability[z][x][y] = true;
+	}
+
+	for(const auto & hero : rightInfo->heroes)
+	{
+		CPathsInfo out(mapSize, hero);
+		auto config = std::make_shared<SingleHeroPathfinderConfig>(out, gameHandler->gameState(), hero);
+		CPathfinder pathfinder(gameHandler->gameState(), config);
+		pathfinder.calculatePaths();
+
+		for (int z = 0; z < mapSize.z; ++z)
+			for (int y = 0; y < mapSize.y; ++y)
+				for (int x = 0; x < mapSize.x; ++x)
+					if (out.getNode({x,y,z})->reachable())
+						rightReachability[z][x][y] = true;
+	}
+
+	for (int z = 0; z < mapSize.z; ++z)
+		for (int y = 0; y < mapSize.y; ++y)
+			for (int x = 0; x < mapSize.x; ++x)
+				if (leftReachability[z][x][y] && rightReachability[z][x][y])
+					return true;
```

Comment:
Can be only one pass thought whole map and single multi array `reachabilityMap`
//store array of paths per hero for left heroes and right heroes
std::vector<CPathsInfo> leftPaths, rightPaths;
for(leftHeroes, righHeroes)
{
leftPaths.emplaceBack(mapSize, hero);
...
}
for(z, y, x)
{
for(leftPaths)
reachabilityMap = true;
for(rightPaths)
if(reachabilityMap)
return true;
}
return false

But not actually critical because as I understand called only once per day

Reply:
True. Wanted to keep code as simple as possible for now since we'll likely need to modify it further based on feedback.

Known issues with current approach:
- Paths per player may be computed multiple times - for each player x player pair. Each player results can be cached instead.
- It does not accounts for other ways player can act - town portal, fly, dimension doors.
- It does not accounts for potential zone guards defeat - you can defeat wandering monsters and attempt to attack player on the other side of the connection.
- And if we go into competitive pvp there are more arcane ways to extend movement range - for example prisons give you hero with max move points.

In most cases however this should not be an issue - this code will not be executed if simturns are inactive or if this check is disabled (not possible in current UI though). I also made adjustments to pathfinder to build only path in 1 turn range instead of entire map to reduce CPU usage. 

And even if player does manage to bypass this check and reaches another player during simturns - all interaction is blocked and should not break the game.

For now I am probably leaving it as it - will need to test & discuss this with MP players. Copying HD Mod approach is not something that can be done in vcmi with low difficulty, so will need feedback from players to decide in which direction to go from here.